### ISSUE TITLE
[20.11.0][fetch] possible memory leak

### ISSUE THREAD

**Issue by xr0master** [nodejs/undici#2846]
### Version

20.11.0

### Platform

Darwin Kernel Version 22.5.0

### What steps will reproduce the bug?

<img width="554" alt="image" src="https://github.com/nodejs/node/assets/8962340/1d2bdcec-9ad1-49de-8a4a-ea5e31425ef0">
<img width="1067" alt="Screenshot 2024-02-21 at 22 38 59" src="https://github.com/nodejs/node/assets/8962340/36165c57-559c-4ad1-a4a5-156e7651ee32">

### What do you see instead?

Each request PerformanceResourceTiming is created

### Additional information

Using express.js

```ts
import express from 'express'; // express server

import type { Application, Request, Response } from 'express';

export const app: Application = express();

app.disable('x-powered-by');

app.get('get', (req: Request, res: Response) => {
  fetch(
    'https://example.com',
  ).then(
    async (response) => {
      const chunks = [] as Uint8Array[];

      if (!response.body) {
        return res.sendStatus(400).end();
      }

      for await (const chunk of response.body) {
        chunks.push(chunk);
      }

      if (response.ok) {
        return res.sendStatus(200).end();
      }

      res.sendStatus(400).end();
    },
    () => {
      res.sendStatus(400).end();
    },
  );
});
```

**Comment by mcollina** [comment]
Does it actually lead to an OOM crash?

**Comment by Ethan-Arrowood** [comment]
I don't believe this is an issue. Fetch is supposed to generate Performance Resource Timings (by spec), and we don't want to automatically clear them in case the user waits to get them until very late in the process. I can't recall if the spec specifies a clean-up or timeout operation for them

**Comment by xr0master** [comment]
@Ethan-Arrowood 
It is not very correct to refer to the spec. There is a noticeable difference in design for the browser and the server.
Getting this issue in the browser is almost impossible since there are few requests per session. However, for the server, this is a highly possible scenario.

**Comment by Ethan-Arrowood** [comment]
Fetch is built to match as closely as possible WHATWG spec. Otherwise it's not fetch anymore. We have attempted to document differences both here and in WinterCG.

**Comment by metcoder95** [comment]
Over a quick research, it seems that it just states that the timing info should be reported after `response is downloaded`, so it seems until fully consumed by the implementor; cannot see or figure out a timing out but maybe until the Response might gc'ed

**Comment by mcollina** [comment]
I think https://github.com/nodejs/undici/pull/1517 would need to be reverted.

The problem is in Node.js. Specifically, it does not expose the `startPerf`, `stopPerf` and `hasObserver` API that it uses internally to implement this without having memory leaks:

https://github.com/nodejs/node/blob/6bb7c4d916e797dfb1624de9f9848cd25193248e/lib/dns.js#L228-L238
https://github.com/nodejs/node/blob/6bb7c4d916e797dfb1624de9f9848cd25193248e/lib/dns.js#L97-L113

**Comment by mcollina** [comment]
Actually, the problem might just be that https://github.com/nodejs/node/blob/6bb7c4d916e797dfb1624de9f9848cd25193248e/lib/internal/perf/resource_timing.js#L208-L230 is enqueining resources even if there aren't observers

**Comment by mcollina** [comment]
I've investigated this further and there is no memory leak for this. The maximum amount of resource timing is by default 250 and can be set via: https://nodejs.org/api/perf_hooks.html#performancesetresourcetimingbuffersizemaxsize.
